File: 1001-2000/1472. Design Browser History.go
package main
// Medium
// Array
// Time:O(1), Space:O(n)
// https://leetcode.cn/problems/design-browser-history/

type BrowserHistory struct {
    history []string
    cur int
}

func Constructor(homepage string) BrowserHistory {
    return BrowserHistory{
        history: []string{homepage},
        cur: 0,
    }
}

func (this *BrowserHistory) Visit(url string)  {
    this.history = this.history[:this.cur + 1]
    this.history = append(this.history, url)
    this.cur++
}

func (this *BrowserHistory) Back(steps int) string {
    this.cur = max(0, this.cur - steps)
    return this.history[this.cur]
}

func (this *BrowserHistory) Forward(steps int) string {
    this.cur = min(len(this.history) - 1, this.cur + steps)
    return this.history[this.cur]
}

File: 1001-2000/1472. Design Browser History.py
class BrowserHistory:

    def __init__(self, homepage: str):
        self.history = [homepage]
        self.cur = 0
    
    # Time:O(k), Space:O(n)
    def visit(self, url: str) -> None:
        self.history = self.history[: self.cur + 1]
        self.history.append(url)
        self.cur += 1
    
    # Time:O(1), Space:O(n)
    def back(self, steps: int) -> str:
        self.cur = max(0, self.cur - steps)
        return self.history[self.cur]
    
    # Time:O(1), Space:O(n)
    def forward(self, steps: int) -> str:
        self.cur = min(len(self.history) - 1, self.cur + steps)
        return self.history[self.cur]
        